impl/src/main/java/org/ehcache/config/builders/ExpiryPolicyBuilder.java

```diff
@@ -64,36 +65,67 @@
     return new TimeToIdleExpiryPolicy(timeToIdle);
   }
 
+  @FunctionalInterface
+  public interface TriFunction<T, U, V, R> {
+    /**
+     * Applies this function to the given arguments.
+     *
+     * @param t the first function argument
+     * @param u the second function argument
+     * @param v the third function argument
+     * @return the function result
+     */
+    R apply(T t, U u, V v);
+  }
+
   /**
    * Fluent API for creating an {@link ExpiryPolicy} instance where you can specify constant values for creation, access and update time.
    * Unspecified values will be set to {@link ExpiryPolicy#INFINITE INFINITE} for create and {@code null} for access and update, matching
    * the {@link #noExpiration()}  no expiration} expiry.
    *
    * @return an {@link ExpiryPolicy} builder
    */
-  public static ExpiryPolicyBuilder expiry() {
-    return new ExpiryPolicyBuilder();
+  public static <K, V> ExpiryPolicyBuilder<K, V> expiry() {
+    return new ExpiryPolicyBuilder<>();
   }
 
-  private Duration create = ExpiryPolicy.INFINITE;
-  private Duration access = null;
-  private Duration update = null;
+  private final BiFunction<? super K, ? super V, Duration> create;
+  private final BiFunction<? super K, ? super Supplier<? extends V>, Duration> access;
+  private final TriFunction<? super K, ? super Supplier<? extends V>, ? super V, Duration> update;
 
-  private ExpiryPolicyBuilder() {}
+  private ExpiryPolicyBuilder() {
+    this((k, v) -> ExpiryPolicy.INFINITE, (k, v) -> null, (k, oldV, newV) -> null);
+  }
+
+  private ExpiryPolicyBuilder(BiFunction<? super K, ? super V, Duration> create,
+                              BiFunction<? super K, ? super Supplier<? extends V>, Duration> access,
+                              TriFunction<? super K, ? super Supplier<? extends V>, ? super V, Duration> update) {
+    this.create = create;
+    this.access = access;
+    this.update = update;
+  }
 
   /**
    * Set TTL since creation
    *
    * @param create TTL since creation
    * @return this builder
```

Comment:
These `@return` stanzas are now wrong... the builder methods now return new builder instances (the builder has become immutable).